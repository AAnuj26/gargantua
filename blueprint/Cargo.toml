[package]
name = "nova"
version = "0.1.0"
edition = "2021"

[dependencies]
derive_setters.workspace = true
serde.workspace = true
serde_json = "1.0.128"
