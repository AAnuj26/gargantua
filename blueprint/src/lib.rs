mod blueprint;
mod parse;
pub use blueprint::*;

pub use url;