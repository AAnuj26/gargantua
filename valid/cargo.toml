[package]
name = "valid"
version = "0.1.0"
edition = "2021"

[dependencies]
derive_setters.workspace = true

[dev-dependencies]
pretty_assertions.workspace = true
