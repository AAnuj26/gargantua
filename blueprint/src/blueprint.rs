use std::collections::{BTreeMap, BTreeSet};

use async_graphql_parser::types::ServiceDocument;
use serde::{Deserialize, Serialize};
use serde_json::Value;
use valid::Valid;

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct Blueprint {
    pub definitions: Vec<Definition>,
    pub schema: SchemaDefinition,
    pub directives: Vec<DirectiveDefinition>,
    pub graphs: Vec<JoinGraph>,
}

#[derive(Clone, Debug, Serialize, Deserialize, Ord, PartialOrd, Eq, PartialEq)]
pub struct GraphId(pub String);

#[derive(Clone, Debug, Serialize, Deserialize, Ord, PartialOrd, Eq, PartialEq)]
pub struct JoinGraph {
    pub name: GraphId,
    pub url: url::Url,
}

impl Blueprint {
    pub fn parse(doc: ServiceDocument) -> Valid<Blueprint, String> {
        super::parse::parse(doc)
    }
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub enum Definition {
    Interface(InterfaceTypeDefinition),
    Object(ObjectTypeDefinition),
    InputObject(InputObjectTypeDefinition),
    Scalar(ScalarTypeDefinition),
    Enum(EnumTypeDefinition),
    Union(UnionTypeDefinition),
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct InterfaceTypeDefinition {
    pub name: String,
    pub fields: Vec<FieldDefinition>,
    pub description: Option<String>,
    pub join_type: Vec<JoinType>,
    pub join_implements: Vec<JoinImplements>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct ObjectTypeDefinition {
    pub name: String,
    pub fields: Vec<FieldDefinition>,
    pub description: Option<String>,
    pub implements: BTreeSet<String>,
    pub join_type: Vec<JoinType>,
    pub join_implements: Vec<JoinImplements>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct InputObjectTypeDefinition {
    pub name: String,
    pub fields: Vec<InputFieldDefinition>,
    pub description: Option<String>,
    pub join_type: Vec<JoinType>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct EnumTypeDefinition {
    pub name: String,
    pub directives: Vec<Directive>,
    pub description: Option<String>,
    pub enum_values: Vec<EnumValueDefinition>,
    pub join_type: Vec<JoinType>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct EnumValueDefinition {
    pub description: Option<String>,
    pub name: String,
    pub directives: Vec<Directive>,
    pub join_enum: Vec<GraphId>
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct SchemaDefinition {
    pub query: Option<String>,
    pub mutation: Option<String>,
    pub subscription: Option<String>,
    pub directives: Vec<Directive>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct InputFieldDefinition {
    pub name: String,
    pub of_type: Type,
    pub default_value: Option<Value>,
    pub description: Option<String>,
    pub join_field: Vec<JoinField>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct FieldDefinition {
    pub name: String,
    pub args: Vec<InputFieldDefinition>,
    pub of_type: Type,
    pub directives: Vec<Directive>,
    pub description: Option<String>,
    pub join_field: Vec<JoinField>
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct Directive {
    pub name: String,
    pub arguments: BTreeMap<String, Value>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct DirectiveDefinition {
    pub name: String,
    pub description: Option<String>,
    pub arguments: Vec<InputFieldDefinition>,
    pub repeatable: bool,
    pub locations: Vec<String>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct ScalarTypeDefinition {
    pub name: String,
    pub directives: Vec<Directive>,
    pub description: Option<String>,
    pub join_type: Vec<JoinType>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct UnionTypeDefinition {
    pub name: String,
    pub directives: Vec<Directive>,
    pub description: Option<String>,
    pub types: BTreeSet<String>,
    pub join_type: Vec<JoinType>,
    pub join_union: Vec<JoinUnion>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct JoinType {
    pub graph: GraphId,
    pub key: Option<String>,
    pub extension: bool,
    pub resolvable: bool,
    pub is_interface_object: bool,
}

impl JoinType {
    pub fn new(graph: String) -> Self {
        Self {
            graph: GraphId(graph),
            key: None,
            extension: false,
            resolvable: true,
            is_interface_object: false,
        }
    }
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct JoinField {
    pub graph: Option<GraphId>,
    pub requires: Option<String>,
    pub provides: Option<String>,
    pub r#type: Option<String>,
    pub external: Option<bool>,
    pub r#override: Option<String>,
    pub used_overridden: Option<bool>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct JoinImplements {
    pub graph: GraphId,
    pub interface: String
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct JoinUnion {
    pub graph: GraphId,
    pub member: String
}

/// Type to represent GraphQL type usage with modifiers
/// [spec](https://spec.graphql.org/October2021/#sec-Wrapping-Types)
#[derive(Clone, Debug, Serialize, Deserialize)]
pub enum Type {
    Named {
        /// Name of the type
        name: String,
        /// Flag to indicate the type is required.
        required: bool,
    },
    List {
        /// Type is a list
        of_type: Box<Type>,
        /// Flag to indicate the type is required.
        non_null: bool,
    },
}
