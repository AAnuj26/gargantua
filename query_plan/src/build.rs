use std::{rc::Rc, vec};

use async_graphql_parser::types as Q;
use blueprint::{Graph, Index};
use valid::Valid;

use crate::SelectionSet;

pub struct Builder<A> {
    index: Rc<Index>,
    _phantom: std::marker::PhantomData<A>,
}

impl<A> Builder<A> {
    pub fn new(index: Rc<Index>) -> Self {
        Self { index, _phantom: std::marker::PhantomData }
    }

    pub fn build(&self, doc: &Q::ExecutableDocument) -> Valid<SelectionSet<A>, String> {
        todo!()
    }
}

#[cfg(test)]
mod test {
    use blueprint::Blueprint;
    use insta::assert_debug_snapshot;
    use resource::resource_str;
    use serde_json::Value;

    use crate::{QueryPlan, SelectionSet};

    #[test]
    fn test() {
        // Blueprint
        let graphql = resource_str!("../fixtures/router.graphql");
        let blueprint = Blueprint::parse(graphql.as_ref()).unwrap();

        // Query
        let query = "query { topProducts { name reviews { score } reviews { description } } }";
        let plan: SelectionSet<Value> =
            SelectionSet::try_new(query.to_string(), &blueprint.to_index()).unwrap();

        assert_debug_snapshot!(plan);
    }
}
