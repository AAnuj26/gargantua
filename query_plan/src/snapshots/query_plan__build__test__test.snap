---
source: query_plan/src/build.rs
expression: plan
---
Fetch {
    service: Graph(
        "Product",
    ),
    query: SelectionSet {
        fields: [],
    },
    representations: None,
    type_name: TypeName(
        "Query",
    ),
}
